src/org/opensolaris/opengrok/web/api/controller/MessagesController.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * See LICENSE.txt included in this distribution for the specific
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2018 Oracle and/or its affiliates. All rights reserved.
+ */
+package org.opensolaris.opengrok.web.api.controller;
+
+import org.opensolaris.opengrok.configuration.RuntimeEnvironment;
+import org.opensolaris.opengrok.web.MessagesContainer;
+import org.opensolaris.opengrok.web.MessagesContainer.Message;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.time.Duration;
+import java.time.format.DateTimeParseException;
+import java.util.Set;
+
+@Path("/messages")
+public class MessagesController {
+
+    private static final String DEFAULT_DURATION = "PT10M";
+    private static final String DEFAULT_TAGS = MessagesContainer.MESSAGES_MAIN_PAGE_TAG;
+    private static final String DEFAULT_CSS_CLASS = "info";
+
+    private final RuntimeEnvironment env = RuntimeEnvironment.getInstance();
+
+    @PUT
+    @Consumes(MediaType.TEXT_PLAIN)
+    public void addMessage(
+            @QueryParam("duration") @DefaultValue(DEFAULT_DURATION) final String durationStr,
```

Comment:
I think this goes against rest design principles. This should be `POST` and the query params should be the body in json.

Reply:
I was actually reading about this because I was not sure and I was not able to find any proper explanation. They say that in body should be data. However, it was not completely clear what they mean by that. I chose this approach only because `duration` and `cssClass` have default values which are not changed often so the ease of use is better here...

However, I have no problem with rewriting it.

Reply:
Query parameters are usef for `GET` queries for filtering and searching.
Here duration and the others are state of the entitity which are being created thus falls into the data in the request body.